File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        // if the tree is empty return empty list
        if (root == nullptr) {
            return {};
        }
        // stack is used to simulate recursion start with the root node
        stack<TreeNode*> stack;
        // store the preorders traversal result
        vector<int> result;
        // push root node 
        stack.push(root);
        // continue until there are no nodes left to process
        while (!stack.empty()) {
            // pop the top node from the stack    
            TreeNode* node = stack.top();
            stack.pop();
            // visit the node (preorder : node -> left -> right)
            result.push_back(node->val);
            // push right child first so the left child is processed first since stack is LIFO
            if (node->right != nullptr) {
                stack.push(node->right);
            }
            // push the left child after right child
            if (node->left != nullptr) {
                stack.push(node->left);
            }
        }
        // return the preorders traversal result
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(H)

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Brute Force & Optimal Code
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # if the tree is empty return empty list
        if not root:
            return []
        # stack is used to simulate start with the root node
        stack = [root]
        # store the preorders traversal result
        result = []
        # continue until there are no nodes left to process
        while stack:
            # pop the top node from the stack
            node = stack.pop()
            #  visit the node (preorder : node -> left -> right)
            result.append(node.val)
            # push right child first so the left child is processed first since stack is LIFO 
            if node.right:
                stack.append(node.right)
            # push the left child after right
            if node.left:
                stack.append(node.left)
        # return the preorders traversal result
        return result

# Time Complexity : O(N)
# Space Complexity : O(1)